### PR TITLE
Template method pattern

### PR DESCRIPTION
**Pull request by PatrickSchiefer**

### CODE REVIEW

File: content/docs/patterns/template-method-pattern/index.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+tags: ["AL", "Interface", "Readability"]
+categories: ["Pattern"]
+---
+
+_Created by Patrick Schiefer, Described by Patrick Schiefer_
+
+## Abstract
+The goal of this pattern is too simplify the solution of  similar problems and make your code more readable
+
+
```
Comment (TheDoubleH) [inline content/docs/patterns/template-method-pattern/index.md:12]:
```suggestion
## Abstract
The goal of this pattern is to simplify the solution of similar problems and make your code more readable.

```

File: content/docs/patterns/template-method-pattern/index.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+## Problem
+In nearly every app you have sometimes to solve similar problems for diferent cases. Mostly not the same developer will solve every case. This results in diferent solutions.
+
```
Comment (TheDoubleH) [inline content/docs/patterns/template-method-pattern/index.md:15]:
```suggestion
## Problem
In nearly every app you sometimes have to solve similar problems for different cases. Mostly not the same developer will solve every case. This results in different solutions.

```

File: content/docs/patterns/template-method-pattern/index.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+## Description
+The pattner is used when you have problems which are independent but require the same logical flow. Examples which occure very often are: Posting Documents, Printing Reports or Exporting Data. 
+
```
Comment (TheDoubleH) [inline content/docs/patterns/template-method-pattern/index.md:18]:
```suggestion
## Description
The pattern is used when you have problems which are independent but require the same logical flow. Examples which occur very often are: _Posting Documents_, _Printing Reports_ or _Exporting Data_. 

```

File: content/docs/patterns/template-method-pattern/index.md [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+Your code gains readability and it is very easy to add new cases for the template. You don't always have to think about the whole logic. You just have to implement the details
+
+## When not to use
+The Pattern should not be used for problems which differ to much. So for example if you have two data exports in your app, one is exporting header and lines and the second one only exports headers. In this case I would suggest to not use the pattern or to make two templates out of it.
+
+
+
+## References 
+[Detailed Explanation of the pattern](https://patrickschiefer.com/2022/04/08/template-method-pattern/)
+
```
Comment (TheDoubleH) [inline content/docs/patterns/template-method-pattern/index.md:157]:
```suggestion
## When not to use
The Pattern should not be used for problems which differ too much. For example, if you have two data exports in your app, one is exporting header and lines and the second one only export header. In this case I would suggest to not use the pattern or to make two templates out of it.

## References 
[Detailed Explanation of the pattern](https://patrickschiefer.com/2022/04/08/template-method-pattern/)

```

**Review by TheDoubleH — COMMENTED** [review]
A Couple of grammar issues

File: content/docs/patterns/template-method-pattern/index.md [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    procedure ExportLine()
+    begin
+        //Generate Exportdata here
+    end;
+
+    procedure NextLine(): Boolean
+    begin
+        exit(SalesLines.Next() > 0);
+    end;
```
Comment (TheDoubleH) [inline content/docs/patterns/template-method-pattern/index.md:117]: Should the template have a Direction (steps)? - possible as an overloaded procedure

```suggestion
    procedure NextLine(Steps : integer): Boolean
    begin
        exit(SalesLines.Next(Steps) <> 0);
    end;
```

**Review by JeremyVyska — APPROVED** [review]

**Review by waldo1001 — APPROVED** [review]

File: content/docs/patterns/template-method-pattern/index.md [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+            WriteToFile();
+          Enum::ExportType::B:
+            SendToWebService();
+        end;
+    end;
+
+    //TODO Implementation of procedures used in example
+}
+```
+As you can see in this example the readability gets bader with every new case.
```
Comment (JesperSchulz) [inline content/docs/patterns/template-method-pattern/index.md:46]: ...readability gets **worse**

File: content/docs/patterns/template-method-pattern/index.md [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+    begin
+        exportImpl.SetSalesHeader(DocType, No);
+        exportInt := exportImpl;
+        export.ExportData(exportInt);
+    end;
+}
+```
+
+## Benefits
+Your code gains readability and it is very easy to add new cases for the template. You don't always have to think about the whole logic. You just have to implement the details
```
Comment (JesperSchulz) [inline content/docs/patterns/template-method-pattern/index.md:147]: Missing period at the end.

File: content/docs/patterns/template-method-pattern/index.md [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+```
+As you can see in this example the readability gets bader with every new case.
+
+## The Pattern
+To implement the Pattern you need at least 3 objects:
+- A template codeunit
+- An Interface which provides the needed procedures
+- A codeunit which Implements the interrface
+
+In my Example I show how to implement a data export with templating
```
Comment (JesperSchulz) [inline content/docs/patterns/template-method-pattern/index.md:54]: Example with capital E?

**Review by JesperSchulz — APPROVED** [review]
LGTM. Only minor comments.

File: content/docs/patterns/template-method-pattern/index.md [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
```
Comment (TheDoubleH) [inline content/docs/patterns/template-method-pattern/index.md:54]: Per @JesperSchulz:

```suggestion
In my example I show how to implement a data export with templating.
```

File: content/docs/patterns/template-method-pattern/index.md (hunk 5 above)
Comment (TheDoubleH) [reply content/docs/patterns/template-method-pattern/index.md]: @JeremyVyska - You are marking all of my comments as resolved without any comment, isn't answering any questions. Please advise.

File: content/docs/patterns/template-method-pattern/index.md [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+    begin
+        exportImpl.SetSalesHeader(DocType, No);
+        exportInt := exportImpl;
+        export.ExportData(exportInt);
+    end;
+}
+```
+
+## Benefits
+Your code gains readability and it is very easy to add new cases for the template. You don't always have to think about the whole logic. You just have to implement the details
```
Comment (TheDoubleH) [inline content/docs/patterns/template-method-pattern/index.md:147]: Per @JesperSchulz:

```suggestion
Your code gains readability and it is very easy to add new cases for the template. You don't always have to think about the whole logic. You just have to implement the details.
```

**Review by TheDoubleH — CHANGES_REQUESTED** [review]

**Review by TheDoubleH — APPROVED** [review]

**Pull request merged**